### PR TITLE
SAK-40938 Put submitted grades in the right format

### PR DESCRIPTION
**Pull request by adrianfish**

### CODE REVIEW

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 1]
```diff
@@ -10854,6 +10854,14 @@ public boolean readGradeForm(RunData data, SessionState state, String gradeOptio
                         }
                     }
                     if (g != null) {
+                        // SAK-40938. Before we stick the grade in the state, make sure it
+                        // has two decimal places.
```
Comment (jonespm) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: This makes sure it has "factor" decimal places, may not be 2.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 2]
```diff
@@ -10854,6 +10854,14 @@ public boolean readGradeForm(RunData data, SessionState state, String gradeOptio
...
+                        String decSeparator = formattedText.getDecimalSeparator();
```
[the remark this answers is not in the document]
Comment (adrianfish) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: I know. The formatting seems to be done in multiple places and needs cleaning up. I can look at that but it'll take some more time. What do you think @ottenhoff ? Have we got time to tidy this formatting up? Basically, this fixes this blocker, but it is layering some more kludge on a big kludge pie, as Matthew politely implies.

[the remark this answers is not in the document]
Comment (adrianfish) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Maybe we put a marker on the refactor for 20. Centralise all the grade formatting.

[the remark this answers is not in the document]
Comment (jonespm) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Huh, kludge pie? So many loc it won't even show up in the github interface? ;) 

It just seems like there's a lot of grade formatting issues that keep coming up between i18n, with how many decimal places to allow, whether to pad the digits or not. That should all be centralized for consistency. But might be more for a future release. I just noticed this new NumberUtil for 19 in  SAK-40563  that started this off.

[the remark this answers is not in the document]
Comment (adrianfish) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: I was more thinking in terms of how many places there seem to be conversion from input to output, adding decimal places or removing them, that kinf of thing. It feels brittle.

**Comment by adrianfish** [comment]
getGradeDisplay is the bit where the the problem 0.00 format is output, depending on input. Unless you provide that with a log10(factor) zero padded number, it gives you the erroneus result. So, we could fix it in the service but that would need testing across a lot of different calls. I didn't want to risk regressions at this stage before the 19.x release so I just fixed the data on the way in.

**Comment by ern** [comment]
Hmm I understand your concern here. If I recall having the service always return a String with a decimal will probably be fine. You could probably double check this?

**Comment by adrianfish** [comment]
Basically, if the service detects an int, 20 say, it will return 0.20 or something similar based on the scale factor. So, to do what I've done with my cleanup on input, I'd have to mod that code to actually append the decimal separator and log10(factor) zeros. Question is, how many places will then be affected by that change? I'll have a go, but I am concerned about causing regressions.

**Comment by adrianfish** [comment]
I moved my code into the service. If you grade an assignment with 40, you get 40.00 back for a scale of 100, which seems to make sense to me - it's what I'd expect back if I was grading. However, the tests fail. My question is: why would we expect to get 0.05 back when we put 5 in as the grade? Have we got a functional spec for assignments somewhere?

**Comment by adrianfish** [comment]
@ern

**Comment by adrianfish** [comment]
I've updated this PR to actually test for the max exceeded alert in the state. If that is set, then we're in review mode and the max has been exceeded. If set, I don't format the grade for display as it should remain as the grader input it in the first place.

**Pull request merged**